### PR TITLE
`asarray`: take the default device into consideration.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106779

Fix: #106773

This PR makes it so `asarray` takes the default device into consideration when called with
a Python sequence as the data.

### CODE REVIEW

**Review by rgommers — APPROVED** [review]
LGTM, thanks for the quick fix @ysiraichi!

**Review by lezcano — APPROVED** [review]

File: test/test_tensor_creation_ops.py [hunk 1]
```diff
@@ -3966,6 +3966,29 @@ def test_numpy_scalars(self, device):
...
         self.assertEqual(tensor.dtype, torch.int32)
 
+    def test_default_device(self, device):
+        import array
+
+        examples = [
+            3,
+            np.arange(5),
+            torch.arange(5),
+            array.array("f", [1, 2, 3, 4]),
```
[the remark this answers is not in the document]
Comment (ysiraichi) [reply test/test_tensor_creation_ops.py:3976]: Ah, you are right. I used `"f"` here because:

**Comment by ysiraichi** [comment]
CI failures are not related.

**Pull request closed without merging**